### PR TITLE
Fix for search failures if shard is in POST_RECOVERY

### PR DESCRIPTION
**Pull request by brwe**
Currently, we do not allow reads on shards which are in POST_RECOVERY which unfortunately can cause search failures on shards which just recovered if there no replicas (#9421). 
The reason why we did not allow reads on shards that are in POST_RECOVERY is that after relocating a shard might miss a refresh if the node that executed the refresh is behind with cluster state processing.  If that happens, a user might execute index/refresh/search but still not find the document that was indexed.
More details [here](https://github.com/elastic/elasticsearch/compare/elastic:master...brwe:replicated_refresh?expand=1#diff-90cb64bedfe055b7d7f1cc28103784d5R971).

@bleskes and I discussed this briefly and he mentioned we could make refresh a replicated operation that goes the same route that index operations go and thereby make sure that the refresh reaches every shard. In this case we could also allow reads on POST_RECOVERY.

I make this PR as a proof of concept so that we can discuss if this is actually a good idea. 
This PR contains:
- a reliable test for #9421
- a fix for #9421 
- a test for the visibility issue that we have when we allow reads in POST_RECOVERY
- the change to make refresh a replicated action just like index, delete, etc.

Let me know what you think. I would make the same changes for flush also.

### CODE REVIEW

**Comment by s1monw** [comment]
I like this a lot! I wonder if we can streamline the implementation here a little more and forward the original request with the shard ID together to have some code we can share between flush and refresh and whatever comes after that? Anyway I think we can just start with what we have here and see how flush turns out afterwards.

**Comment by brwe** [comment]
I was actually wondering if we should also have a dedicated Action, similar to TransportBroadcastAction something like TransportReplicatedBroadcastAction or something that flush and refresh can derive from.

File: core/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportRefreshAction.java [hunk 1]
```diff
@@ -46,74 +51,91 @@
 /**
  * Refresh action.
  */
-public class TransportRefreshAction extends TransportBroadcastAction<RefreshRequest, RefreshResponse, ShardRefreshRequest, ShardRefreshResponse> {
+public class TransportRefreshAction extends HandledTransportAction<RefreshRequest, RefreshResponse> {
 
-    private final IndicesService indicesService;
+    private final TransportReplicatedRefreshAction replicatedRefreshAction;
+    ClusterService clusterService;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportRefreshAction.java]: can this be private final?

File: core/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportRefreshAction.java [hunk 2]
```diff
@@ -46,74 +51,91 @@
...
+                                  ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver, TransportReplicatedRefreshAction replicatedRefreshAction) {
+        super(settings, RefreshAction.NAME, threadPool, transportService, actionFilters, indexNameExpressionResolver, RefreshRequest.class);
+        this.replicatedRefreshAction = replicatedRefreshAction;
+        this.clusterService = clusterService;
     }
 
     @Override
-    protected RefreshResponse newResponse(RefreshRequest request, AtomicReferenceArray shardsResponses, ClusterState clusterState) {
+    protected void doExecute(final RefreshRequest request, final ActionListener<RefreshResponse> listener) {
+        logger.trace("execute refresh");
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportRefreshAction.java]: is this trace needed?

File: core/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportRefreshAction.java [hunk 3]
```diff
@@ -46,74 +51,91 @@
...
+        GroupShardsIterator groupShardsIterator = clusterService.operationRouting().searchShards(clusterService.state(), indexNameExpressionResolver.concreteIndices(clusterService.state(), request), null, null);
+        final AtomicReferenceArray<ReplicatedRefreshResponse> shardsResponses = new AtomicReferenceArray(groupShardsIterator.size());
+        final AtomicInteger numPrimariesToProcess = new AtomicInteger(groupShardsIterator.size());
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportRefreshAction.java]: we have a `CountDown` class that helps to do this and prevents you from counting down more than actually specified...

File: core/src/main/java/org/elasticsearch/action/replicatedrefresh/ReplicatedRefreshRequest.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    public ShardId getShardId() {
+        return shardId;
+    }
+
+    public void setShardId(ShardId shardId) {
+        this.shardId = shardId;
+        index(shardId.index().name());
+    }
+
+    ShardId shardId;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/replicatedrefresh/ReplicatedRefreshRequest.java]: can this be private?

**Comment by brwe** [comment]
I continued on this. I tried to generalize what I did for refresh so that it can be used for flush too. Now I wonder: should this work for synced flush too?

**Comment by s1monw** [comment]
@brwe what's the status here, do you wait for reviews?

**Comment by brwe** [comment]
@s1monw yes. @bleskes might have an opinion on that too.

**Comment by bleskes** [comment]
@brwe and I talked this through and we decided to try and simplify things and remove some intermediate abstractions. Concretely try to:

1) Use ReplicationRequest/ReplicationResponse/TransportReplicationAction directly rather than having ReplicatedBroadcastShardRequest/ReplicatedBroadcastShardResponse/TransportReplicatedBroadcastShardAction.
2) Use BroadcastRequest/BroadcastResponse instead of ReplicatedBroadcastRequest/ReplicatedBroadcastResponse
3) Rename TransportReplicatedBroadcastAction to TransportBroadcastReplicationAction   

Also, we should break this into two PRs:
1) One changing the refresh/flush behavior.
2) A follow up PR to change the read on POST_RECOVERY semantics.

**Comment by brwe** [comment]
> Also, we should break this into two PRs:
> 1) One changing the refresh/flush behavior.
> 2) A follow up PR to change the read on POST_RECOVERY semantics.

I made a pr for the first part here: #13068

**Comment by brwe** [comment]
Opened the second pr for the actual fix now here: https://github.com/elastic/elasticsearch/pull/13246 
I'll close this one here now.

**Pull request closed without merging**